### ISSUE TITLE
fill confusion when subtracting in some cases with lines coinciding

### ISSUE THREAD

**Issue by mlwilkerson** [Logicalshift/flo_curves#23]
I haven't yet understood exactly what is the cause of this problem, but it does seem like one factor involved is that the shape being subtracted has a line that coincides with the base shape--the front edge of the dog's leg--which seems to result in fill-rule confusion: what ends up being "inside" versus "outside".

However, when I tried to create a more minimal test case, simulating merely that scenario of such lines overlapping, it did not reproduce the unexpected results. So there may be something else going on here.

This is based on a Font Awesome Pro icon, `dog` in the Classic Regular style. I've redacted it a bit for licensing reasons, just to show the problem.

Here's the base shape:

```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512">
<!--! Font Awesome Pro 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2023 Fonticons, Inc. -->
<path d="M318 342.2C329.4 351.3 336 365.1 336 379.6V448C336 456.8 343.2 464 352 464H368C376.8 464 384 456.8 384 448V274.3L300.5 223L318 342.2zM335.1 0 C340.7 0 346 2.6 349.4 7.1L360.6 22 364 26.6 364.2 26.9L368 32H444.1C456.8 32 469 37.1 478 46.1L496 64H528C554.5 64 576 85.5 576 112V144C576 197 533 240 480 240H434.7L432 256V448C432 483.3 403.3 512 368 512H352C316.7 512 288 483.3 288 448V428.6 379.6L250 200L335.1 0zM336.9 189L390.3 221.8C397.5 204.1 414.9 192 434.7 192H480C506.5 192 528 170.5 528 144V112H496C483.3 112 471.1 106.9 462.1 97.9L444.1 80 368 80C363.8 80 359.6 79.4 355.6 78.4L338 183.9C337.7 185.6 337.3 187.3 336.9 189zM416 112A16 16 0 1 1 448 112 16 16 0 1 1 416 112z"/></svg>
```

Looks like this, rendered:
![original](https://user-images.githubusercontent.com/11575183/231256111-a32bd357-fb45-4858-99aa-9531c148e411.png)

I don't know how significant it is, but the original SVG also uses the default nonzero fill rule. I noticed in the code comments that the flo_curves subtract method uses the evenodd rule. In all cases, I simply convert the flo_curve results back into nonzero fill-rule and it works as expected. In this case, however, I still get unexpected results. So I'm not sure whether that's a significant factor here--but I mention it just in case. It isn't the _only_ significant factor anyway, since my other cases work fine when I convert to nonzero fill-rule after the path subtraction.

Here's the shape that is being subtracted:
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><path d="M384 352V480c0 11.7 3.1 22.6 8.6 32H640V296.6 272C640 210.1 589.9 160 528 160S416 210.1 416 272v24.6C396.9 307.6 384 328.3 384 352z"/></svg>
```

Looks similar to this (though this screenshot may not be scaled correctly, the SVG source should be correct):
<img width="462" alt="cutout" src="https://user-images.githubusercontent.com/11575183/231257687-b32ba5a8-355b-4d8d-9651-baf52b55a849.png">

After subtracting (with or without conversion to nonzero fill-rule), it looks unexpectedly like this:
![rendered-after-subtract](https://user-images.githubusercontent.com/11575183/231258196-bde97f0d-12d8-4a02-9327-012694552264.png)

The result SVG:
```svg
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><path d="M300.5 223L318 342.2C329.4 351.3 336 365.1 336 379.6C336 402.4 336 425.2 336 448C336 456.8 343.2 464 352 464C357.3 464 362.7 464 368 464C376.8 464 384 456.8 384 448C384 416 384 384 384 352C384 326.1 384 300.2 384 274.3C356.2 257.2 328.3 240.1 300.5 223zM416 112C416 120.8 423.2 128 432 128C440.8 128 448 120.8 448 112C448 103.2 440.8 96 432 96C423.2 96 416 103.2 416 112z"/></svg>
```

If I were to nudge the cutout shape to the right by `1.0`, it looks closer to the expected result, in that it looks like the cutout shape has simply been subtracted from the original shape, without the fill confusion. (It's _not_ the actual expected result, of course--translating to the right a bit just seems to avoid the overlap of the line on the left side where the fill confusion seems to be occurring):
![rendered-subtract-after-nudging-right](https://user-images.githubusercontent.com/11575183/231258534-ce090319-8cea-4013-9601-30490e1af887.png)

I've compared this to the subtract operation in [Skia's PathOps subtract](https://www.npmjs.com/package/pathkit-wasm), and Skia produces the expected result.

**Comment by mlwilkerson** [comment]
If memory serves, I've reproduced this problem both on the v0.7 and v0.8 branches of flo_curves.

**Comment by Logicalshift** [comment]
Hm, I need to do a bit more work to be sure but I'm fairly sure this is a reappearance of an old bug: there are a pair of intersecting collinear lines and I think one of the intersections is not being added to the result. I think this will be `overlapping_region()` returning the wrong results most likely.

There are some other possibilities: collisions are found using a left-to-right scanning algorithm that could potentially be fooled by a vertical line (with the same result), and there are a few ways that the ray casting algorithm could get confused.

If you're interested, there's a function `graph_path_svg_string()` that writes out the intermediate structure and makes it possible to see these bugs a bit more clearly: for subtracting it can be used with the output of the following part of `path_sub()`:

```
    let mut merged_path = GraphPath::new();
    merged_path         = merged_path.merge(GraphPath::from_merged_paths(path1.iter().map(|path| (path, PathLabel(0)))));

    merged_path         = merged_path.collide(GraphPath::from_merged_paths(path2.iter().map(|path| (path, PathLabel(1)))), accuracy);
    merged_path.round(accuracy);

    merged_path.set_exterior_by_subtracting();
```

(It'll illustrate all of the edges that were found and which it decided were external or internal, and can also be given a list of rays to show what the ray caster thought it saw)

**Comment by Logicalshift** [comment]
Here's the output of the debugging function for a variant of this case: I think this confirms my first suspicion on close inspection. The point labelled '5 <- 4 - 18' should be two points for where the two overlapping lines intersect at the end, and because it's not there's a break in the path (you can drag the lines around in an editor to see this more clearly). I think this is a case where the lines are intersecting while going in opposite directions, so the two intersection points are at the end points of both lines, except one has been missed out.

Another bug is a bit more subtle: the blue edges are those found to be exterior and one of the two edges making up the overlapping line is marked this way. It shouldn't be, as the ray here is passing from outside to outside. I think this is probably an artifact of this same bug right now (the code that deals with this expects overlaps to be identical so won't work right after an incorrect collision has been generated)

![debug_5](https://user-images.githubusercontent.com/4084632/231290767-e39ff85c-43e1-4707-a24e-0a63946a34fb.svg)

**Comment by Logicalshift** [comment]
I think I've fixed this now in the just released v0.7.1. I think problem was indeed just that overlapping lines moving in the opposite direction weren't colliding properly: the new version fixes this and adds some tests for as many variants of overlapping lines as I could think of.

**Comment by mlwilkerson** [comment]
Wow, awesome! I have confirmed the fix in v0.7.1.
